### PR TITLE
Clip slice range expressions

### LINKED ISSUE #52: `slice` corner case: Index range at runtime that is beyond a tensor does not return a zero-element tensor
The `slice` implementation inside nvFuser does not have a runtime check to determine if the slice range is beyond the dimension size it is slice.  Pytorch returns a zero-element tensor while nvFuser returns an empty tensor of the size of the slice(s).

Example:

```
import torch
from nvfuser import FusionDefinition, DataType

acts = [
    torch.randn(5, 5, device='cuda'),
]

def legal(fd: FusionDefinition) -> None :
    T0 = fd.from_pytorch(acts[0])
    T1 = fd.ops.slice(T0, start_indices=[6, 6], end_indices=[8, 8], strides=[1, 1])
    fd.add_output(T1)

with FusionDefinition() as fd:
    legal(fd)

out = fd.execute(acts)
print(out[0].size(), out[0].stride())

out_eager = acts[0][6:8, 6:8]
print(out_eager.size(), out_eager.stride())
```

Output:
```
$ python test.py 
tensor([[0., 0.],
        [0., 0.]], device='cuda:0')
torch.Size([2, 2]) (2, 1)
tensor([], device='cuda:0', size=(0, 0))
torch.Size([0, 0]) (5, 1)
```

Naoya's comments on the feasibility of a runtime check:

> I think we would need some runtime check. Currently we are ignoring a patten like:
```
t0: [I0]
t1 = slice(t0, {{0, 1}})
t1: [I1]
```
>Here, the size of I1 is just 1, so it should be marked as a broadcast domain 
>But at the codegen side, we are not doing this
>In this particular case, this should be trivial as we can obviously find the extent is 1, but more generally, whether the extent is 1 or not depends the extent of the slided domain as well as the start and end offsets
>So, we need to do some check at the run time using the runtime info.
>And that also seems to be the above case

### LINKED ISSUE #439: Slice should clip to end of tensor
Slice behavior currently does not match numpy/PyTorch when the `end_indices` extend past the tensor extent:
```c++
// Slice with end beyond size of input. This should clip to input, not pad.
TEST_F(NVFuserTest, FusionResizeSlice6_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  std::vector<int64_t> shape({9});

  // concrete shapes to avoid dynamic Fusion
  auto tv0 = makeConcreteTensor(shape);
  fusion.addInput(tv0);

  auto tv1 = slice(tv0, {{fusion.zeroVal(), IrBuilder::create<Int>(11)}});
  fusion.addOutput(tv1);

  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);

  auto t0 = at::randn(shape, options);
  std::vector<c10::IValue> aten_inputs({t0});

  FusionExecutor fe;
  fe.compileFusion(&fusion, aten_inputs);
  auto cg_outputs = fe.runFusion(aten_inputs);

  auto ref = t0.index({at::indexing::Slice(0, 11)});

  testValidate(&fusion, cg_outputs, aten_inputs, {ref}, __LINE__, __FILE__);
  // C++ exception with description "The size of tensor a (9) must match the
  // size of tensor b (11) at non-singleton dimension 0 Exception raised from
  // infer_size_impl at /opt/pytorch/pytorch/aten/src/ATen/ExpandUtils.cpp:31
}
```
In these cases, we should clip to the end of the tensor. That is, we should implicitly use `min(extent, end_index)` instead of `index` itself. As discussed in #397 this means we should not concretize slice ops prematurely unless we know their input extents.

### PR DESCRIPTION
**Pull request by jacobhinkle**
This PR normalizes the inputs to `slice` in order to mimic the semantics of numpy/PyTorch slicing. For an axis with extent `ext`, if we receive a slice of `(start, stop, step)` we normalize it to `(norm_start, norm_stop, step)` where
```
norm_start = max(0, start < 0 ? start + ext : start);
norm_stop = max(norm_start, min(ext, stop < 0 ? stop + ext : stop));
```

Specific changes in this PR:
- Form the above expressions in the `slice` op.
- Add shmoo tests that test various scenarios with constant and input size slices.

The simple Fusion in the input range test prints like this:
```
Inputs:
  T0_g[ iS0{9} ], float
  i3, nvfuser_index_t
  i4, nvfuser_index_t
Outputs:
  T1_g[ ?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf ], float

%kernel_math {
b7 = i3 < 0;
i5 = i3 + 9;
i9 = where(b7, i5, i3);
i11 = fmax(0, i9);
b15 = i4 < 0;
i13 = i4 + 9;
i17 = where(b15, i13, i4);
i19 = fmin(9, i17);
i21 = fmax(i11, i19);
T1_g[ ?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf ]
   = slice( T0_g[ iS0{9} ], { {i11, i21, 1} } )
}

T0_g[ iS0{9} ]
 root domain : (iS0{9})
 contiguity: f
 leaf domain : (iS0{9})
T1_g[ ?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf ]
 root domain : (iS1{9}rf)
  Resize: iS1{9}rf by ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) and ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) -> ?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf
 rfactor domain : (?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf)
 contiguity: t
 leaf domain : (?S2{( ( ( -( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ) ) + 9 ) + ( ( fmax(( fmax(0, ( where(( i3 < 0 ), ( i3 + 9 ), i3) )) ), ( fmin(9, ( where(( i4 < 0 ), ( i4 + 9 ), i4) )) )) ) - 9 ) )}rf)
```
resulting in the following CUDA kernel:
```c++
__global__ void kernel1(Tensor<float, 1, 1> T0, nvfuser_index_t i0, nvfuser_index_t i1, Tensor<float, 1, 1> T1) {
  nvfuser_index_t i2;
  i2 = i0 + 9;
  bool b3;
  b3 = i0 < 0;
  nvfuser_index_t i4;
  i4 = b3 ? i2 : i0;
  nvfuser_index_t i5;
  i5 = max(0, i4);
  nvfuser_index_t i6;
  i6 = i1 + 9;
  bool b7;
  b7 = i1 < 0;
  nvfuser_index_t i8;
  i8 = b7 ? i6 : i1;
  nvfuser_index_t i9;
  i9 = min(9, i8);
  nvfuser_index_t i10;
  i10 = max(i5, i9);
  nvfuser_index_t i11;
  i11 = (-i5) + i10;
  nvfuser_index_t i12;
  i12 = i5 * T0.alloc_stride[0];
  #pragma unroll 1
  for(nvfuser_index_t i13 = 0; i13 < i11; ++i13) {
    T1[i13]
       = T0[(i12 + (T0.alloc_stride[0] * i13))];
  }
}
```

This PR does NOT simplify these expressions for non-constant inputs. This can be done at concretization, which will be left for a follow-up PR.

Stacked on #892 and #895.

Fixes #439. Fixes #52.

### CODE REVIEW

**Comment by jacobhinkle** [comment]
!build

File: csrc/ops/alias.cpp [hunk 1]
```diff
@@ -691,12 +691,18 @@ TensorView* slice(TensorView* inp, const std::vector<Slice>& ranges) {
       out_root_id = inp_root_id->cloneWithoutRFactor();
       out_rf_id = out_root_id;
     } else {
+      // Clip the start and stop values to the extent of the input
+      auto clipped_start =
```
Comment (naoyam) [inline csrc/ops/alias.cpp:695]: What about negative values? Should they be clipped to zero?

Comment (jacobhinkle) [reply csrc/ops/alias.cpp:695]: Ordinarily we would add extent to negative values, which we can do with a where expression.

Comment (naoyam) [reply csrc/ops/alias.cpp:695]: What would happen if it's still negative?

Comment (jacobhinkle) [reply csrc/ops/alias.cpp:707]: If it's negative at runtime we will wind up padding I think, which is also incorrect. I pushed a change just now that adds `where()` to handle negative start/stop. However, it leads to an error when trying to simplify the resulting index expressions. We can handle constant negative values by translating them at definition time, but that would still mean that dynamic `Val`s that take on negative values would lead to incorrect outputs.

Comment (jacobhinkle) [reply csrc/ops/alias.cpp:707]: Example output:
```
%kernel_math {
b5 = 11 < 0;
i4 = 11 + 9;
i6 = where(b5, i4, 11);
b8 = 13 < 0;
i7 = 13 + 9;
i9 = where(b8, i7, 13);
T1_g[ iS2{( ( -( fmin(9, ( where(( 11 < 0 ), ( 11 + 9 ), 11) )) ) ) + ( fmin(9, ( where(( 13 < 0 ), ( 13 + 9 ), 13) )) ) )}rf ]
   = slice( T0_g[ iS0{9} ], { {i6, i9, 1} } )
}
```

File: test/test_resize.cpp [hunk 2]
```diff
@@ -1122,6 +1122,76 @@ TEST_F(NVFuserTest, FusionResizeSlice5_CUDA) {
   testValidate(&fusion, cg_outputs, aten_inputs, {t2, t4}, __LINE__, __FILE__);
 }
 
+// Test slice with a variety of (constant) inputs
+TEST_F(NVFuserTest, FusionResizeSliceShmoo_CUDA) {
+  for (auto [start, stop] : std::vector<std::pair<int64_t, int64_t>>(
+           {// Slice with end beyond size of input. This should clip to input,
+            // not pad.
+            {0, 11},
+            {11, 13},
```
Comment (jacobhinkle) [inline test/test_resize.cpp:1131]: This case blows the stack with infinite recursion trying to simplify index expression. The first case works fine.

Comment (naoyam) [reply test/test_resize.cpp:1131]: Sounds like a bug of `SimplifyingIrBuilder`

File: csrc/ops/alias.cpp (hunk 1 above)
Comment (jacobhinkle) [reply csrc/ops/alias.cpp:707]: > What would happen if it's still negative?

I will check in the AM but I think in pytorch it's an error to provide a value < -extent.

Comment (jacobhinkle) [reply csrc/ops/alias.cpp:707]: In both torch and numpy, it seems that providing a negative value first adds `extent` then clips to `>= 0`. Values are also clipped to `<= extent`, so any int is valid input for either start or stop and the expressions to use for `a` equal to either start or stop are:
```python
normed_a = min(max(where(a < 0, extent + a, a), 0), extent)
```
These will be pretty tough to simplify if we don't enforce which branch we're on. We can do that at concretization by determining the branch as one of:
```c++
class enum SliceIndexBranch {
  AlwaysZero, // a <= -extent
  Negative, // -ext < a < 0
  Positive, // 0 <= a < extent
  AlwaysExtent // extent <= a
};
```
These map to the following much simpler expressions
```
AlwaysZero => zeroVal(),
Negative => extent() + a,
Positive => a,
AlwaysExtent => extent()
```
I think doing this at concretization makes sense because users commonly write slicing code with only one or two of these branches in mind, so we should not try and compile a kernel that handles them all. For example, a user might hardcode an upper limit that could be either `Positive` or `AlwaysExtent`, but they would not ordinarily mix `Negative` and `Positive` in the same code I would think. Even if they do, by concretizing we'll have faster kernels, and the only price paid is a recompile when we see a new pattern.

File: test/test_resize.cpp (hunk 2 above)
Comment (jacobhinkle) [reply test/test_resize.cpp:1131]: I think it's just due to not handling the `where` expression I implemented. We could probably get it to terminate, but it won't be able to simplify it since we can't prove which branch we've taken unless we concretize differently, in which case we should probably just go ahead and replace the expression with its known simpler form (see above).

**Comment by jacobhinkle** [comment]
Closing in favor of #511.

File: test/test_resize.cpp [hunk 3]
```diff
@@ -2360,7 +2427,7 @@ TEST_F(ResizeTest, Slice1DVectorizeManual1) {
   FusionGuard fg(fusion_ptr.get());
 
   const int64_t slice_offset = 4;
-  const std::vector<int64_t> shape({1024 * 1024});
+  const std::vector<int64_t> shape({1024L * 1024L});
```
Comment (jacobhinkle) [inline test/test_resize.cpp:2430]: Silencing clang-tidy

File: csrc/ir/builder.cpp [hunk 4]
```diff
@@ -330,6 +330,9 @@ Val* SimplifyingIrBuilder::addExpr(Val* lhs, Val* rhs) {
 }
 
 Val* SimplifyingIrBuilder::subExpr(Val* lhs, Val* rhs) {
+  if (lhs->sameAs(rhs)) {
+    return lhs->fusion()->zeroVal(lhs->dtype());
+  }
```
Comment (jacobhinkle) [inline csrc/ir/builder.cpp:335]: This lets us hardcode zero extent when normalized stop == normalized start.

File: csrc/ir/builder.cpp [hunk 5]
```diff
@@ -373,6 +376,9 @@ Val* SimplifyingIrBuilder::divExpr(Val* lhs, Val* rhs) {
   if (rhs->isOneInt()) {
     return lhs;
   }
+  if (lhs->sameAs(rhs)) {
+    return lhs->fusion()->oneVal(lhs->dtype());
+  }
```
Comment (jacobhinkle) [inline csrc/ir/builder.cpp:381]: In same vein as above, but unused in this PR

File: csrc/kernel_cache.cpp [hunk 6]
```diff
@@ -895,8 +895,8 @@ FusionKernelRuntime::FusionKernelRuntime(
       fusion.get());
 
   if (isDebugDumpEnabled(DebugDumpOption::FusionIrPreseg)) {
-    std::cout << "Fusion IR after pre-segmenter optimization passes:"
-              << std::endl;
+    debug() << "Fusion IR after pre-segmenter optimization passes:"
+            << std::endl;
```
Comment (jacobhinkle) [inline csrc/kernel_cache.cpp:899]: Unrelated to this PR. Just found wrong ostream in this debug dump.

File: csrc/evaluator_common.cpp [hunk 7]
```diff
@@ -353,6 +353,14 @@ NaiveValueMachine::NaiveValueMachine(PrecomputedValues& precomputed_values)
...
       } else if (auto bop = dynamic_cast<BinaryOp*>(def)) {
         makeBinaryOp(bop);
+      } else if (auto top = dynamic_cast<TernaryOp*>(def)) {
+        makeTernaryOp(top);
+      } else if (auto lsop = dynamic_cast<LoadStoreOp*>(def)) {
+        NVF_ERROR(
+            lsop->opType() == LoadStoreOpType::Set,
+            "NaiveValueMachine: unsupported LoadStoreOpType: ",
+            lsop->opType());
+        makeSetOp(lsop);
```
Comment (jacobhinkle) [inline csrc/evaluator_common.cpp:363]: Unused in this PR currently but we already had `SET_OP`, so this lets along with `makeSetOp` us evaluate `set`.

**Comment by jacobhinkle** [comment]
!build

File: csrc/kernel_cache.cpp [hunk 8]
```diff
@@ -897,8 +897,8 @@ FusionKernelRuntime::FusionKernelRuntime(
       fusion.get());
 
   if (isDebugDumpEnabled(DebugDumpOption::FusionIrPreseg)) {
-    std::cout << "Fusion IR after pre-segmenter optimization passes:"
-              << std::endl;
+    debug() << "Fusion IR after pre-segmenter optimization passes:"
```
Comment (jacobhinkle) [inline csrc/kernel_cache.cpp:900]: Unrelated to this PR. Just found this incorrect use of `std::cout`.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
The test `ResizeTest.FusionResizeMultiSliceEmpty` now fails due to #610.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

File: csrc/ops/alias.cpp [hunk 9]
```diff
@@ -704,36 +701,58 @@ TensorView* slice(TensorView* inp, const std::vector<Slice>& ranges) {
...
       ndims);
 
-  auto normalize_slice_range = [](Slice range, Val* extent) -> Slice {
+  const auto normalize_slice_range = [](Slice range, Val* extent) -> Slice {
+    auto cast_extent =
+        SimplifyingIrBuilder::maybeCastExpr(DataType::Index, extent);
+
+    auto zero = FusionGuard::getCurFusion()->zeroVal(DataType::Index);
+
+    // norm_start = max(0, start < 0 ? start + extent : start)
```
Comment (zasdfgbnm) [inline csrc/ops/alias.cpp:710]: What if start > extent? In python, this is valid:
```python
[1,2,4][10000:]
```
is
```python
[]
```

Comment (zasdfgbnm) [reply csrc/ops/alias.cpp:710]: Oh, sorry, looks like this case is already tested.

File: test/test_resize.cpp [hunk 10]
```diff
@@ -1124,6 +1124,109 @@ TEST_F(ResizeTest, FusionResizeSlice5) {
...
+    auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+    auto t0 = at::randn(shape, options);
+    std::vector<c10::IValue> aten_inputs({t0});
+
+    FusionExecutor fe;
+    fe.compileFusion(&fusion, aten_inputs);
+    auto cg_outputs = fe.runFusion(aten_inputs);
+
+    auto ref = t0.index({at::indexing::Slice(start, stop)});
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1159]: Do we still need to compute `ref` manually? I think Priya's work should cover this.

File: test/test_resize.cpp [hunk 11]
```diff
@@ -1124,6 +1124,109 @@ TEST_F(ResizeTest, FusionResizeSlice5) {
...
+            {-11, 0}})) {
+    std::vector<c10::IValue> aten_inputs({t0, start, stop});
+    auto cg_outputs = fe.runFusion(aten_inputs);
+
+    auto ref = t0.index({at::indexing::Slice(start, stop)});
+
+    testValidate(&fusion, cg_outputs, aten_inputs, {ref}, __LINE__, __FILE__);
+  }
+}
+
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1229]: Should we have a test for the same shape and set of slices, but directly feed it into `FusionExecutorCache`?

File: csrc/ops/alias.cpp (hunk 9 above)
Comment (jacobhinkle) [reply csrc/ops/alias.cpp:710]: Yes, I wrote it this way to avoid an additional ternary expression, instead of clipping start to `extent`. If `start > extent`, then `stop` will get clipped to `start`, since it is first clipped to `extent` then to `start`.

File: test/test_resize.cpp [hunk 12]
```diff
@@ -1124,6 +1124,109 @@ TEST_F(ResizeTest, FusionResizeSlice5) {
...
+            {3, 4},
+            {7, 5},
+            {0, 11},
+            {11, 13},
+            {-3, 8},
+            {-3, -1},
+            {-3, -5},
+            {13, -1},
+            {-11, 9},
+            {-11, 0}})) {
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1220]: Could you also test {-13, -11} where both start and stop are out of range to the left?

File: test/test_resize.cpp [hunk 13]
```diff
@@ -1124,6 +1124,109 @@ TEST_F(ResizeTest, FusionResizeSlice5) {
   testValidate(&fusion, cg_outputs, aten_inputs, {t2, t4}, __LINE__, __FILE__);
 }
 
+// Test slice with a variety of constant ranges
+TEST_F(NVFuserTest, FusionResizeSliceConstantShmoo_CUDA) {
+  for (auto [start, stop] : std::vector<std::pair<int64_t, int64_t>>(
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1129]: Should we use the same set of slices as `FusionResizeSliceInputShmoo_CUDA`?

File: test/test_resize.cpp (hunk 11 above)
Comment (jacobhinkle) [reply test/test_resize.cpp:1229]: Yes that's a good idea, since in the above I am manually concretizing as Iteration before running all slices. I'll add that.

File: csrc/dynamic_transform.cpp [hunk 14]
```diff
@@ -316,7 +316,7 @@ void DynamicTransformConcretizationInfo::analyzeResizes(
         out_id->toString());
     auto extent_int = extent_val.as<int64_t>();
     NVF_ERROR(
-        extent_int > 0,
+        extent_int >= 0,
```
Comment (jacobhinkle) [inline csrc/dynamic_transform.cpp:319]: Good idea to test `FusionExecutorCache` @zasdfgbnm. This was exposed when I added that test. I think we didn't hit this in other empty tensor tests since they didn't include only a single empty output without empty intermediates. Anyway, I think this fix makes sense. After this all the slice tests are passing.

File: test/test_resize.cpp [hunk 15]
```diff
@@ -1124,6 +1124,154 @@ TEST_F(ResizeTest, FusionResizeSlice5) {
...
+    DynamicTransform::concretizeFusion(&fusion, &info);
+    NVF_CHECK(
+        !fusion.hasDynamicTransform(), "Expected to have no dynamic transform");
+  }
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion);
+
+  auto t0 = at::randn(shape, options);
+  for (auto [start, stop] : std::vector<std::pair<int64_t, int64_t>>(
```
Comment (zasdfgbnm) [inline test/test_resize.cpp:1204]: Should we pull this set of slices out of the test and reuse it for all the three tests?

File: test/test_resize.cpp (hunk 13 above)
Comment (jacobhinkle) [reply test/test_resize.cpp:1129]: Yes that's good now. The reason I didn't originally is just that it slows down the test a lot since we need to recompile for each slice.

File: test/test_resize.cpp (hunk 15 above)
Comment (jacobhinkle) [reply test/test_resize.cpp]: Done

**Review by zasdfgbnm — APPROVED** [review]

**Review by naoyam — APPROVED** [review]
LGTM
